### PR TITLE
[DO NOT MERGE] [SQL] Add Managed Instance DNS alias cmdlets

### PR DESCRIPTION
**Pull request by krivi37**
## Description

Added commands for managing Managed Instance DNS aliases: New, Get, Set, Move, Remove.
Please **do not merge** yet, we need only a review.

## Checklist

- [X] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [X] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [X] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1215
    - [X] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.ps1 [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    This function accepts DNS alias object and expected parameters. If the $hasAzureDnsRecord is passed, then we check that the dns record is different from null.
+    If the expectedResourceId is passed, we also check that field (this is only used for set operations so we check if it stayed the same after update).
+#>
+function ValidateDnsAlias($managedInstanceDnsAliasActualObject, $expectedAliasName, $expectedRGName, $expectMIName, $hasAzureDnsRecord, $expectedResourceId)
+{
+    Assert-AreEqual $expectedAliasName $managedInstanceDnsAliasActualObject.DnsAliasName
+    Assert-AreEqual $expectedRGName $managedInstanceDnsAliasActualObject.ResourceGroupName
+    Assert-AreEqual $expectMIName $managedInstanceDnsAliasActualObject.ManagedInstanceName
+    Assert-NotNull $managedInstanceDnsAliasActualObject.Id
+    Assert-Null $managedInstanceDnsAliasActualObject.PublicDnsRecord
```
Comment (ostojicb) [inline src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.ps1:36]: out of curiosity, why is this always null?

File: src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿using Microsoft.Azure.Commands.ScenarioTest.SqlTests;
```
Comment (ostojicb) [inline src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.cs:1]: nitpick: you're missing microsoft headers in some files

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/GetAzureSqlManagedInstanceDnsAlias.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        [ValidateNotNullOrEmpty]
+        public string InstanceName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the name of the Azure Sql Managed Instance DNS Alias
+        /// </summary>
+        [Parameter(Mandatory = false, ParameterSetName = GetByNameParameterSet, Position = 2, HelpMessage = "Name of the managed instance DNS alias.")]
+        [Parameter(Mandatory = false, ParameterSetName = GetByParentObjectParameterSet, Position = 1, HelpMessage = "Name of the managed instance DNS alias.")]
+        [Alias("DnsAliasName")]
+        [SupportsWildcards]
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/GetAzureSqlManagedInstanceDnsAlias.cs:47]: how many aliases can there be per instance?

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        public SwitchParameter AsJob { get; set; }
+
+        protected override IEnumerable<AzureSqlManagedInstanceDnsAliasModel> GetEntity()
+        {
+            // Since we didn't construct source DNS alias resource Id, for this parameter set, in the ExecuteCmdlet method,
+            // we can do it now, since ModelAdapter is initialized at this point (it was initialized after entering the base.ExecuteCmdlet method).
+            if(ParameterSetName == MoveByNamesParameterSet)
+            {
+                ConstructSourceResourceId(SourceResourceGroupName, SourceInstanceName, SourceName, ModelAdapter.Context.Subscription.Id);
+            }
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs:135]: is this required?

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        private void ConstructSourceResourceId(AzureSqlManagedInstanceDnsAliasModel sourceDnsAlias)
+        {
+            SourceResourceId = sourceDnsAlias.Id;
+            SourceName = sourceDnsAlias.DnsAliasName;
+        }
+
+        private void ParseParentObjectParameters()
+        {
+            ResourceGroupName = InstanceObject.ResourceGroupName;
+            InstanceName = InstanceObject.ManagedInstanceName;
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs:168]: nitpick, these kind of 2liner voids with sideeffects are not a great practice IMO (when sideeffect is not required).
Also we're not really checking for the 'input' here, e.g. the latter one kinda prays we'll have `InstanceObject` when invoked which is not really great (think broader than this PR, i.e. copy pasting, refactoring, lack of tests etc) + I don't think you actually typed less code this way 😄

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/RemoveAzureSqlManagedInstanceDnsAlias.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+using System.Globalization;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Sql.ManagedInstanceDnsAlias.Cmdlet
+{
+    [Cmdlet(VerbsCommon.Remove, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceDnsAlias",
+        DefaultParameterSetName = DeleteByNameParameterSet,
+        SupportsShouldProcess = true),
+        OutputType(typeof(bool))]
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/RemoveAzureSqlManagedInstanceDnsAlias.cs:16]: i think in your tests there's a comment like:
```
 #Since -PassThru was passed, we expect that the remove cmdlet returned the original object
 ```
 that sounds a bit contradictory to the specified `OutputType`, can you double check what's up?

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/GetAzureSqlManagedInstanceDnsAlias.cs (hunk 3 above)
Comment (krivi37) [reply src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/GetAzureSqlManagedInstanceDnsAlias.cs:47]: I'm not sure if a limit exists, but there can be more than one

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs (hunk 5 above)
Comment (krivi37) [reply src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs:168]: Yeah, they should have some checks running at least. This Move command class had more parameter sets initially so I used this to avoid copy pasting for multiple parameter sets, but I ommited refactoring this when I removed some of the unnecessary parameter sets.
TLDR: I will inline those :)

File: src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+# Move-AzSqlInstanceDnsAlias
+
+## SYNOPSIS
+Modifies the managed instance to which Azure SQL Managed Instance DNS Alias is pointing.
+
+## SYNTAX
+
+### MoveByNameAndSourceResourceIdParameterSet (Default)
+```
+Move-AzSqlInstanceDnsAlias [-ResourceGroupName] <String> [-InstanceName] <String> [-SourceResourceId] <String>
```
[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: It is the current resource ID of the DNS alias (old one). After execution of the Move command, that same alias will point to the target managed instance.

File: src/Sql/Sql/help/Set-AzSqlInstanceDnsAlias.md [hunk 8]
```diff
@@ -0,0 +1,304 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Sql.dll-Help.xml
+Module Name: Az.Sql
+online version:
+schema: 2.0.0
+---
+
+# Set-AzSqlInstanceDnsAlias
```
[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Set-AzSqlInstanceDnsAlias.md:8]: This is a PUT request

File: src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (BethanyZhou) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: So the ResourcegroupName and InstanceName here is used for target managed instance.

[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: That's right

[the remark this answers is not in the document]
Comment (BethanyZhou) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: I'm wondering whether it would be more friendly to customer if we use parameter ResourceId to name current resource ID and use   DestinationResourceGroupName and DestinationInstanceName to name ResourcegroupName and InstanceName? We have some similar cmdlets with this pattern. See [Move-AzResource](https://docs.microsoft.com/en-us/powershell/module/az.resources/move-azresource?view=azps-8.1.0), [Move-AzDataLakeGen2Item](https://docs.microsoft.com/en-us/powershell/module/az.storage/move-azdatalakegen2item?view=azps-8.1.0), ...

[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: There is a problem here. All SQL cmdlets inherit from AzureSqlCmdletBase, and there exists a mandatory parameter called ResourceGroupName (that is why it is overridden here in the code). If I change base class, it will induce breaking changes in other cmdlets. But I could add the alias for the parameter. What do you think about adding the alias?

File: src/Sql/Sql/help/Get-AzSqlInstanceDnsAlias.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+```
+
+## DESCRIPTION
+Get the specific Azure SQL Managed Instance DNS Alias or lists all Managed Instance DNS Aliases for the managed instance.
+
+## EXAMPLES
+
+### Example 1: Get a specific managed instance DNS alias
+```powershell
+PS C:\> Get-AzSqlInstanceDnsAlias -ResourceGroupName <resourceGroupName> -InstanceName <managedInstanceName> -Name <dnsAliasName>
```
[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Get-AzSqlInstanceDnsAlias.md]: Fixed

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs (hunk 5 above)
Comment (krivi37) [reply src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs]: I've inlined those functions

File: src/Sql/Sql/help/Get-AzSqlInstanceDnsAlias.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Sql.dll-Help.xml
+Module Name: Az.Sql
+online version:
```
[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Get-AzSqlInstanceDnsAlias.md]: Added

File: src/Sql/Sql/help/Get-AzSqlInstanceDnsAlias.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+Get the specific Azure SQL Managed Instance DNS Alias or lists all Managed Instance DNS Aliases for the managed instance.
+
+## EXAMPLES
+
+### Example 1: Get a specific managed instance DNS alias
+```powershell
+PS C:\> Get-AzSqlInstanceDnsAlias -ResourceGroupName <resourceGroupName> -InstanceName <managedInstanceName> -Name <dnsAliasName>
+
+ResourceGroupName    : <rgName>
+ManagedInstanceName  : <managedInstanceName>
```
[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Get-AzSqlInstanceDnsAlias.md:45]: Added

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs (hunk 4 above)
Comment (krivi37) [reply src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs]: It is, but I've inlined it now

File: src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (BethanyZhou) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: I think you can hide the ResourceGroupName from base class by `private new string ResourceGroupName { get; set; }`, could you have a try?

[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: It cannot be hidden that way- it won't not compile because of https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0114?f1url=%3FappId%3Droslyn%26k%3Dk(CS0114) which is treated as an error here.

It can be overridden by specifying public override string [Mandatory = false] ResourceGroupName, and then declaring a new mandatory parameter DestResourceGroupName.
However, I am not sure if this is a good practice, because we declare an unused field, just in order to hide it. Wouldn't declaring an alias for ResourceGroupName be descriptive enough?

[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: Will leave the alias then :)

[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: Hi, I tried again to hide the property and it worked. It seems I ommited the "new" keyword here. I will check how it looks now, and will update the PR. Thanks for the suggestion!

[the remark this answers is not in the document]
Comment (BethanyZhou) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md:17]: Please let me know this PR is ready to review by adding label `needs-review`

[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Move-AzSqlInstanceDnsAlias.md]: I've pushed new commits and added the label, please review :)

File: src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+using Microsoft.Azure.Commands.ScenarioTest.SqlTests;
+using Microsoft.WindowsAzure.Commands.ScenarioTest;
+using Xunit;
+using Xunit.Abstractions;
+using RestTestFramework = Microsoft.Rest.ClientRuntime.Azure.TestFramework;
+
+namespace Microsoft.Azure.Commands.Sql.Test.ScenarioTests
+{
+    public class ManagedInstanceDnsAliasTests : SqlTestsBase
```
Comment (BethanyZhou) [inline src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.cs:23]: Our test framework changes, please pull latest code from main branch and follow latest test writing styple.

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        /// Gets or sets the name of destination managed instance.
+        /// </summary>
+        [Parameter(Mandatory = true, ParameterSetName = MoveByNamesParameterSet, Position = 1, HelpMessage = "Name of the destination managed instance.")]
+        [Parameter(Mandatory = true, ParameterSetName = MoveByNameAndSourceParentObjectParameterSet, Position = 1, HelpMessage = "Name of the destination managed instance.")]
+        [Parameter(Mandatory = true, ParameterSetName = MoveByNameAndSourceInputObjectParameterSet, Position = 1, HelpMessage = "Name of the destination managed instance.")]
+        [Parameter(Mandatory = true, ParameterSetName = MoveByNameAndSourceResourceIdParameterSet, Position = 1, HelpMessage = "Name of the destination managed instance.")]
+        [ResourceNameCompleter("Microsoft.Sql/managedInstances", nameof(AzureSqlManagedInstanceDnsAliasCmdletBase.ResourceGroupName))]
+        [ValidateNotNullOrEmpty]
+        [Alias("DestInstanceName")]
+        public string InstanceName { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs:73]: Please use DestInstanceName as its formal name.

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        /// <summary>
+        /// Gets or sets the instance Object
+        /// </summary>
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = MoveByParentObjectAndSourceNameParameterSet, Position = 0, HelpMessage = "Input object of the destination managed instance.")]
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = MoveByParentObjectsParameterSet, Position = 0, HelpMessage = "Input object of the destination managed instance.")]
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = MoveByParentObjectAndSourceInputObjectParameterSet, Position = 0, HelpMessage = "Input object of the destination managed instance.")]
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = MoveByParentObjectAndSourceResourceIdParameterSet, Position = 0, HelpMessage = "Input object of the destination managed instance.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("DestInstanceObject")]
+        public AzureSqlManagedInstanceModel InstanceObject { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs:84]: Please use DestInstanceObject as its formal name.

File: src/Sql/Sql/help/Set-AzSqlInstanceDnsAlias.md [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+## OUTPUTS
+
+### Microsoft.Azure.Commands.Sql.ManagedInstanceDnsAlias.Model.AzureSqlManagedInstanceDnsAliasModel
+
+## NOTES
+
+## RELATED LINKS
+
+https://aka.ms/sqlmi-dnsalias-docs
```
Comment (BethanyZhou) [inline src/Sql/Sql/help/Set-AzSqlInstanceDnsAlias.md:316]: See to add related cmdlet link https://github.com/Azure/azure-powershell/blob/c230bcf4fa8417d28df1617c29142477dbeb052a/src/Accounts/Accounts/help/Get-AzContext.md?plain=1#L137

File: src/Sql/Sql/help/Remove-AzSqlInstanceDnsAlias.md [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+## DESCRIPTION
+Removes a specific Azure SQL Managed Instance DNS Alias.
+
+## EXAMPLES
+
+### Example 1: Removes a managed instance DNS alias
+```powershell
+Remove-AzSqlInstanceDnsAlias -ResourceGroupName ResourceGroup1 -InstanceName ManagedInstance1 -Name DnsAlias1
+```
+
```
[the remark this answers is not in the document]
Comment (krivi37) [reply src/Sql/Sql/help/Remove-AzSqlInstanceDnsAlias.md:48]: Added

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs (hunk 13 above)
Comment (krivi37) [reply src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs]: Changed

File: src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.cs (hunk 12 above)
Comment (krivi37) [reply src/Sql/Sql.Test/ScenarioTests/ManagedInstanceDnsAliasTests.cs]: Changed

File: src/Sql/Sql/help/Set-AzSqlInstanceDnsAlias.md (hunk 15 above)
Comment (krivi37) [reply src/Sql/Sql/help/Set-AzSqlInstanceDnsAlias.md]: Added

File: src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs (hunk 14 above)
Comment (krivi37) [reply src/Sql/Sql/ManagedInstanceDnsAlias/Cmdlet/MoveSqlManagedInstanceDnsAlias.cs]: Changed

**Comment by isra-fel** [comment]
@krivi37 Hello, the pull request looks good to us. However I noticed you've added the "Do not merge" label. What's your plan on this PR?

**Comment by krivi37** [comment]
> @krivi37 Hello, the pull request looks good to us. However I noticed you've added the "Do not merge" label. What's your plan on this PR?

Hi @isra-fel , the plan was to target the September release, but because of some backend issues, we will postpone it for October. Should I convert the PR to draft, and reactivate it when it is ready to merge?

**Comment by isra-fel** [comment]
Please do convert it to draft. Thanks! @krivi37

**Comment by NoriZC** [comment]
Closing the PR for lack of activity. Feel free to reopen to continue working on it. Thanks 😀
